sympy/solvers/solveset.py

```diff
@@ -1718,10 +1723,30 @@ def linear_eq_to_matrix(equations, *symbols):
         # Extract coeff of symbols
         coeff_list = []
         for symbol in symbols:
-            coeff_list.append(f.coeff(symbol))
+            coeff_list.append(f.diff(symbol))
```

Comment:
You can check the linearity now by seeing `if check_linearity and coeff_list[-1].free_symbols & set(symbols): raise Error`.

If this passes, you have a linear coefficient of an equation which was linear or had a removable singularity (e.g. f(x) = x*(1/x + 1)` has a singularity at f(0)`. You can make future derivatives faster by replacing symbol with 0. If you get S.NaN there was a removable singularity and if you want to ignore it, don't do the replacement.

```
newf = f.xreplace({symbol: 0})
if newf is not S.NaN:
    f = newf
```

Reply:
...or better, expand this portion of the expression to get rid of the singularity:

```
newf = f.xreplace({symbol: 0})
if newf is S.NaN:
    i, d = f.as_independent(symbol)
    f = f.func(i, expand_mul(d)).xreplace({symbol: 0})
```